Azure CLI access

Tener opened the pull request:
Fixes https://github.com/gravitational/teleport/issues/16401

Setup:
1. Deploy Teleport instance to Azure VM. 
2. Ensure the VM has access to at least one [user-assigned identity](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/qs-configure-portal-windows-vm#user-assigned-managed-identity). Make sure that your user-assigned identity has access to some resources. 
3. Configure Azure as an application. `cloud: Azure` is the critical part of the config.
```yaml
app_service:
  enabled: "yes"
  apps:
  - name: my_azure
    cloud: Azure
```
4. Ensure the user is allowed to use chosen identities. If this is a fresh cluster, the `access` role should contain the snippet below. If not, modify a role the user has access to include reference to `{{internal.azure_identities}}`. You can also hardcode the list of identities in the role itself (not recommended).
```yaml
spec:
  allow:
    azure_identities:
    - '{{internal.azure_identities}}'
```
5. When using `{{internal.azure_identities}}` the list of allowed roles is sourced from user-specific traits. To update existing user use `tctl users update <username> --set-azure-identities ID_1,ID_2,ID_3,...`, where `ID_N` is full name of Azure identity (e.g. `/subscriptions/0a2a4eaa-0764-4a8e-bd55-5ae521b6ee6f/resourceGroups/RESOURCE_GROUP/providers/Microsoft.ManagedIdentity/userAssignedIdentities/IDENTITY_NAME` )
6. Refresh client login with `tsh logout; tsh login --proxy=...`.
7. Log in to `my_azure` app with `tsh app login my_azure [--azure-identity IDENTITY_NAME]`. Providing an identity name is optional if there is just one available. The identity can be provided in shortened form (e.g. only IDENTITY_NAME suffix) or as a full identifier (required if IDENTITY_NAME is ambiguous). 
8. Run commands with `tsh az ...`, e.g. `tsh az vm list`. 
9. Alternatively, start a proxy with `tsh proxy azure` and follow printed instructions.

Tener committed:
Azure CLI access

Tener committed:
Check error.

Tener committed:
Tests, proper JWT signing

Tener committed:
Merge branch 'master' into tener/azure-cli

Tener committed:
Lint fixes.

Tener committed:
Add missing license.

Tener committed:
Remove testing code.

ptgott commented:
One thing that tripped me up a bit while testing this was that items in the `azure_identities` field within a role's allow/deny conditions need to use the full URI of the identity, not just the name—is there a way to add validation for this when a user creates a role?

ptgott commented on tool/tsh/azure.go:
```diff
@@ -0,0 +1,423 @@
+	}
+
+	// suffix match?
+	expectedSuffix := fmt.Sprintf("/Microsoft.ManagedIdentity/userAssignedIdentities/%v", cf.AzureIdentity)
```
If the user is copying the identity resource ID from the JSON view (the link on the right in the screenshot below), the resource ID will be lowercased (e.g., `/subscriptions/0000000/resourcegroups/paulgottschling/providers/microsoft.managedidentity/userassignedidentities/teleport-az`). Should this function do a case-insensitive comparison?

<img width="1254" alt="image" src="https://user-images.githubusercontent.com/25250668/204902412-9dc70d4a-2417-422a-a13e-e9b6ad8bbf15.png">

Tener replied:
@ptgott I've made changes both to `tsh app login` (so that "teleport-az" matches "/subscriptions/0000000/resourcegroups/paulgottschling/providers/microsoft.managedidentity/userassignedidentities/Teleport-AZ" ) and to RBAC engine (mainly to ensure that "Allow: foo" gets denied by "Deny: FOO").

greedy52 reviewed (commented):
first pass

Tener commented on api/types/app.go:
```diff
@@ -296,9 +304,21 @@ func (a *AppV3) CheckAndSetDefaults() error {
+		a.Spec.Cloud = CloudAWS
+	}
+	switch a.Spec.Cloud {
+	case "", CloudAWS, CloudAzure, CloudGCP:
```
It isn't supported right now, true, but it is next in line to be worked on.

greedy52 replied:
ok, I see it's one of Q4 goals. If it doesn't make it though, it's better to return an error until it's supported.

Tener commented on api/utils/azure/endpoints.go:
```diff
@@ -24,6 +24,42 @@ import (
+	}
+
+	for _, suffix := range suffixes {
+		if strings.HasSuffix(hostname, suffix) {
```
Good catch!

Tener commented on api/types/app_test.go:
```diff
@@ -198,14 +199,20 @@ func TestAppIsAWSConsole(t *testing.T) {
 			assertIsAWSConsole: require.False,
 		},
+		{
+			name:               "CLI-only AWS App",
+			cloud:              CloudAWS,
+			assertIsAWSConsole: require.True,
```
Probably, yeah, although it should work as is for API-only access.

Tener committed:
Properly check hostname suffixes

Tener commented on lib/client/api.go:
```diff
@@ -2734,6 +2738,23 @@ func (tc *TeleportClient) CreateAppSession(ctx context.Context, req types.Create
+	ctx, span := tc.Tracer.Start(
+		ctx,
+		"teleportClient/GetAppSession",
+		oteltrace.WithSpanKind(oteltrace.SpanKindClient),
```
What is the guidance on what should be recorded in attributes? I did a quick scan of methods in `api.go`, but I'm having trouble inferring a consistent rule.

greedy52 replied:
@rosstimothy do we have guide lines on this?

Tener committed:
Fix docs, trace session.

Tener committed:
Fix docs.

Tener commented on lib/srv/app/azure/handler.go:
```diff
@@ -0,0 +1,298 @@
+	return resp, nil
+}
+
+func (s *Forwarder) formatForwardResponseError(rw http.ResponseWriter, r *http.Request, err error) {
```
Yeah, we will have an analogous problem here with Azure (and GCP next).

I wonder if we can attach this error message to a custom response header (`x-teleport-api-error`?) and sniff that header on `tsh` side?

greedy52 replied:
I think that's a great idea.

Tener replied:
My first attempt: https://github.com/gravitational/teleport/pull/18675/commits/8af456faf46f0d049f41dba37111c6bfb4a4a5c9

```
$ tsh proxy azure --port 3500
Started Azure proxy on http://127.0.0.1:3500.

Use the following credentials and HTTPS proxy setting to connect to the proxy:

  export AZURE_CONFIG_DIR=/Users/tener/.azure-teleport-azure
  export HTTPS_PROXY=http://127.0.0.1:3500
  export MSI_ENDPOINT=https://azure-msi.teleport.dev/my-own-secret
  export REQUESTS_CA_BUNDLE=/Users/tener/.tsh/keys/boson.tener.io/tener-app/boson.tener.io/azure-localca.pem

[2022/12/06 13:47:58] local_proxy.go:269 [Warn] ERROR REPORT: Original Error: context.deadlineExceededError context deadline exceeded Stack Trace:
	github.com/gravitational/teleport/lib/srv/app/azure/handler.go:249 github.com/gravitational/teleport/lib/srv/app/azure.(*Forwarder).replaceAuthHeaders
	github.com/gravitational/teleport/lib/srv/app/azure/handler.go:205 github.com/gravitational/teleport/lib/srv/app/azure.(*Forwarder).prepareForwardRequest
	github.com/gravitational/teleport/lib/srv/app/azure/handler.go:151 github.com/gravitational/teleport/lib/srv/app/azure.(*Forwarder).RoundTrip
	github.com/gravitational/oxy@v0.0.0-20221029012416-9fbf4c444680/forward/fwd.go:174 github.com/gravitational/oxy/forward.(*httpForwarder).serveHTTP
	github.com/gravitational/oxy@v0.0.0-20221029012416-9fbf4c444680/forward/fwd.go:167 github.com/gravitational/oxy/forward.(*Forwarder).ServeHTTP
	github.com/gravitational/teleport/lib/srv/app/server.go:853 github.com/gravitational/teleport/lib/srv/app.(*Server).serveSession
	github.com/gravitational/teleport/lib/srv/app/server.go:815 github.com/gravitational/teleport/lib/srv/app.(*Server).serveHTTP
	github.com/gravitational/teleport/lib/srv/app/server.go:778 github.com/gravitational/teleport/lib/srv/app.(*Server).ServeHTTP
	github.com/gravitational/teleport/lib/auth/middleware.go:617 github.com/gravitational/teleport/lib/auth.(*Middleware).ServeHTTP
	github.com/gravitational/teleport/lib/httplib/httplib.go:93 github.com/gravitational/teleport/lib/httplib.MakeTracingHandler.func1
	net/http/server.go:2109 net/http.HandlerFunc.ServeHTTP
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.36.4/handler.go:204 go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*Handler).ServeHTTP
	net/http/server.go:2947 net/http.serverHandler.ServeHTTP
	net/http/server.go:1991 net/http.(*conn).serve
	runtime/asm_arm64.s:1172 runtime.goexit

	User Message: context deadline exceeded
```

I'm slightly worried this may be too much information for the end user, though. What do you think?

greedy52 replied:
Thanks a lot for addressing this!

> I'm slightly worried this may be too much information for the end user, though.

Personally, I think just the error message is good enough. The client-side user cannot do much with the stack trace of the server anyway. That said I don't mind having it either.

Either way, I think we should be clear that this error comes from the server's response.

smallinsky replied:
I'm having mixed feelings about exposing stack and internal teleport error context to the tsh user. 
Most likely it won't bring any value to the user because `ERROR REPORT: Original Error: context.deadlineExceededError context deadline exceeded Stack Trace:` doesn't mean anything to the `tsh az` user that doesn't know Teleport/GO context. Also, there is a risk that the debug log will leak some kind of internal information about teleport or Azure. Though I agree that  `x-teleport-api-error` is a mechanism to return the error to the user but we should have more control over what we are returning and be more verbose in error message.

Tener replied:
@smallinsky 

I was following https://github.com/gravitational/teleport/issues/16571 here, where it notes that the fallback to diagnosing the errors is checking the service logs:

> What would you like Teleport to do? Improve the AWS CLI proxy (`tsh aws`)'s error logging. At the moment, when a command fails we aren't getting anything useful back, instead we get `error 500 when calling XYZ operation`. We are still able to find the real error by checking `Teleport` service logs on the app proxy node which shows the actual error: `[AWS:SIGNE] Failed to process request: Post "https://rds.us-west-1c.amazonaws.com/": dial tcp: lookup [rds.us-west-1c.amazonaws.com](http://rds.us-west-1c.amazonaws.com/) on 17x.10.0.2:5x: no such host`.

What is your proposal to address https://github.com/gravitational/teleport/issues/16571? Return only the `err.Error()` string? Something else?

smallinsky replied:
I think that we should implement more suited mechanism for error propagation instead of forwarding the server error to the aws/azure cli client. 

For instance we can use a user error and in case of server error create a descriptive user CLI error on the server and forward it using `x-teleport-api-error` mechanism. But after a second thought we can start with current solution and alter this approach.

Tener commented on lib/srv/app/azure/handler.go:
```diff
@@ -0,0 +1,298 @@
+func (s *Forwarder) replaceAuthHeaders(r *http.Request, sessionCtx *common.SessionContext, reqCopy *http.Request) error {
+	auth := reqCopy.Header.Get("Authorization")
+	if auth == "" {
+		s.Log.Debugf("No Authorization header present, skipping replacement.")
+		return nil
```
Yep, this is a valid thing you can do if the API in question doesn't require authentication. There are few like that. If this is an error to omit `Authorization` then the Azure endpoint will tell you this.

We could route `Authorization`-less requests entirely outside of Teleport, but it feels like an unnecessary complexity. I don't see the benefit of doing so.

greedy52 replied:
sounds good. I was just curious. I think it's better this way so we can have it in our audit log.

Tener committed:
Update comments for azureApp proxies.

Tener commented on tool/tsh/azure.go:
```diff
@@ -0,0 +1,423 @@
+	// Set proxy settings.
+	if a.localForwardProxy != nil {
+		envVars["HTTPS_PROXY"] = "http://" + a.localForwardProxy.GetAddr()
+		envVars["HTTP_PROXY"] = "http://" + a.localForwardProxy.GetAddr()
```
Good point, this neither works nor is needed.

Tener commented on tool/tsh/azure.go:
```diff
@@ -0,0 +1,423 @@
+		// 1. `tsh az login` in one console
+		// 2. `az ...` in another console
+		// without custom config dir the second invocation will hang, attempting to connect to (inaccessible without configuration) MSI.
+		"AZURE_CONFIG_DIR": path.Join(homeDir, ".azure-teleport-"+a.app.Name),
```
I don't think we should remove these files even if they are stale. Among other things az CLI stores user-provided configuration (see: https://learn.microsoft.com/en-us/cli/azure/azure-cli-configuration). If we were to wipe out that folder on logout it would naturally be lost.

For the same reason I don't think we should use `~/.tsh` as a parent directory.

smallinsky reviewed (commented):
First pass

Tener commented on lib/srv/alpnproxy/azure_msi_middleware.go:
```diff
@@ -0,0 +1,184 @@
+	return nil
+}
+
+func (m *AzureMSIMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
```
Currently `AzureMSIMiddleware` is only concerned with minting new tokens as requested.

What kind of local validation would you see here?

Tener replied:
> @Tener
> What would you say about leveraging the `MSI_ENDPOINT` env variable for that ? For instance we can generate a uuid token and instead of using `https://azure-msi.teleport.dev`as MSI_ENDPOINT we canst the <random_uuid>msi.teleport.dev as MSI_ENDPOINT and verify if a client used proper value inside local proxy ?

We could in principle set MSI_ENDPOINT to `https://azure-msi.teleport.dev/<secure hex>`. This would be annoying in some scenarios though, for the same reason random `--port` in the `tsh proxy azure`. You need to propagate the configuration to third-party tools and if the configuration changes on each restart, this is a bad UX. But maybe we can mitigate this in similar fashion to `--port` flag, i.e. by allowing the secret to be read from env variable?

greedy52 replied:
> Right now we have several services that works with tunnel mode without local validate mode.

Sounds good.

Another reason I brought up this topic as I recall this issue #16229 an edge case for AWS, not sure this would apply to Azure though.

So say user runs `tsh app login` as AWS role A and launched `tsh proxy aws` for terraform. Terraform may use role A to assume an role B for certain operations. Obvisouly our local proxy gets requests signed for both role A and role B. We currently error out for requests signed for role B but ideally we can just route traffic for role B directly to AWS without forwarding to Teleport backend.

Tener replied:
We now set the `MSI_ENDPOINT` to `https://azure-msi.teleport.dev/<random hex>`, unless the user has `MSI_ENDPOINT` already set to `https://azure-msi.teleport.dev/<user chosen secret>` in which case we will use the provided secret.

smallinsky commented on api/types/app.go:
```diff
@@ -296,9 +304,21 @@ func (a *AppV3) CheckAndSetDefaults() error {
 	if a.Spec.URI == "" {
-		return trace.BadParameter("app %q URI is empty", a.GetName())
+		if a.Spec.Cloud != "" {
+			a.Spec.URI = fmt.Sprintf("cloud://%v", a.Spec.Cloud)
```
Why we need artificially set the URI if we have the a.spec.Cloud property  ?

Tener replied:
I'm pretty sure we make an assumption in other places in code that URI is never empty. This is a reasonable assumption to make, so instead of invalidating it we introduce a new scheme instead. We have an existing precedent for this approach: TCP apps which employ `tcp://...` scheme.

smallinsky reviewed (commented):
First pass

Tener commented on lib/jwt/jwt.go:
```diff
@@ -317,6 +335,25 @@ func (k *Key) VerifySnowflake(p SnowflakeVerifyParams) (*Claims, error) {
+
+	// Validate the signature on the JWT token.
+	var out AzureTokenClaims
+	if err := tok.Claims(k.config.PublicKey, &out); err != nil {
+		return nil, trace.Wrap(err)
+	}
```
> I'm getting following error during testing:

Can you make sure you are using the latest code everywhere (e.g. `make full`)? This error sounds like `tsh` using the old code for minting the auth token.

Tener replied:
@smallinsky unless the message above is truncated this is old code; current code has different message:

https://github.com/gravitational/teleport/blob/94e0db6c160be7d5fe543e9ac6cd4ef29512d388/tool/tsh/app.go#L141-L143

smallinsky replied:
👍  Thanks I have been using stale tsh build

Tener committed:
Address review comments:
- add secret to MSI_ENDPOINT url
- `tsh` will respect MSI_ENDPOINT if present, allowing to pre-set the secret
- `tsh app login` picks Azure identities in case-insensitive way

Tener committed:
RBAC changes for Azure Identities:
- wildcard "*" matcher is respected, but only in "deny"
- matches are done in case-insenstiive way between allow.azure_identities and deny.azure_identities
- if Azure identity is invalid, a warning is logged to server logs
- `tctl users add`/`tctl users update` commands reject invalid Azure identities

Tener commented:
@ptgott 

> > One thing that tripped me up a bit while testing this was that items in the `azure_identities` field within a role's allow/deny conditions need to use the full URI of the identity, not just the name—is there a way to add validation for this when a user creates a role?
> 
> This is a good point from a usability perspective. I'll see what I can do about this.

This turned out trickier than I was hoping for, but I did make some changes. The difficulty comes from the fact that there is no single place we can add this validation to. In the end I've added:
- a warning in case Azure Identity looks invalid - this is to be found in server logs;
- an error if you try to set an invalid identity with `tctl users add` or `tctl users update`.

Tener committed:
LocalProxyHTTPMiddleware requires CheckAndSetDefaults() method

Tener committed:
Cache strings.ToLower(identity)

Tener committed:
Drop invalid test

Tener committed:
Merge branch 'master' into tener/azure-cli

Tener committed:
Merge branch 'master' into tener/azure-cli

Tener committed:
Set timeout for acquiring access token.

Tener committed:
Pass underlying App error to tsh via X-Teleport-API-Error header.

greedy52 commented on tool/tsh/azure.go:
```diff
@@ -0,0 +1,457 @@
+
+	sort.Strings(identities)
+
+	t := asciitable.MakeTable([]string{"Available Azure identities"}, identities)
```
seems this would only print one identity. The `MakeTable` expects `(headers, row1, row2, row3)` so this way it treats `identities` as a single row but it only prints out one column.

Tener replied:
Good catch, thanks!

Tener committed:
Fix printing of multiple Azure identities.

smallinsky commented on lib/web/apps.go:
```diff
@@ -357,15 +358,20 @@ func (h *Handler) proxyDNSNames() (dnsNames []string) {
 	for _, server := range appServers {
+		a := server.GetApp()
+		// Skip over API-only Cloud apps
+		if strings.HasPrefix(a.GetURI(), "cloud://") {
```
what if the Cloud is `AWS` added by [a.URI = fmt.Sprintf("cloud://%v", a.Cloud)](https://github.com/gravitational/teleport/pull/18675/files#diff-054e31280a78f4aea4759cd740f8d8bcecf563679d7fe6b79991fdb944e8c611R1113) where CLI and Console access uses the same app ?

Tener replied:
If the Teleport admin wants to expose the AWS console as App, then they must set the URI to console for the chosen instance of AWS (public/gov cloud/China). Otherwise, the interpretation is that they don't want it to be exposed.

This is open to interpretation, and we may want to change how this detail works in the future. For existing configurations things work exactly as before.

greedy52 commented on tool/tsh/azure.go:
```diff
@@ -0,0 +1,465 @@
+	return envVars, nil
+}
+
+// GetForwardProxyAddr returns local forward proxy address.
+func (a *azureApp) GetForwardProxyAddr() string {
+	if a.localForwardProxy != nil {
+		return a.localForwardProxy.GetAddr()
+	}
+	return ""
+}
+
+// GetEndpointURL returns Azure endpoint URL that clients can use.
+func (a *azureApp) GetEndpointURL() string {
+	if a.localALPNProxy != nil {
+		return "https://" + a.localALPNProxy.GetAddr()
+	}
+	return ""
+}
```
nit: not used

greedy52 reviewed (approved):
Tested today with an Azure VM. Works as expected. 

Curious is there anyway use the managed identity outside Azure (e.g. on AWS ec2)? 

Anyway. Overall looks good to me.

Minor nitpicks:
1. It would be great we  can add `--cloud` to `teleport app start`, or maybe update the description for url with an `cloud://` example
2. Could be just me, that after using `tsh az xxx`, i was expecting the proxy version to be `tsh proxy az` but it's actually `tsh proxy azure`. not sure if worth to have an alias

Tener committed:
Address review comments.

Tener commented on lib/srv/app/azure/handler.go:
```diff
@@ -0,0 +1,303 @@
+func (s *Forwarder) getToken(ctx context.Context, managedIdentity string, scope string) (*azcore.AccessToken, error) {
+	key := cacheKey{managedIdentity, scope}
+
+	ctxT, cancel := context.WithTimeout(ctx, time.Second*5)
```
Done.

Tener commented on lib/srv/app/azure/handler.go:
```diff
@@ -0,0 +1,303 @@
+			reqCopy.Header.Add(key, v)
+		}
+	}
+	reqCopy.Header.Del("Content-Length")
```
You are right, this isn't needed in this case.

Tener committed:
Add `trace.Wrap` as needed.

Tener committed:
Address ParseResourceID feedback

Tener commented on lib/services/role.go:
```diff
@@ -291,6 +292,24 @@ func filterInvalidWindowsLogins(candidates []string) []string {
+	}
+}
+
+var azureIdentityPattern = regexp.MustCompile(`(?i)^/subscriptions/([a-fA-F0-9-]+)/resourceGroups/([0-9a-zA-Z-]+)/providers/Microsoft\.ManagedIdentity/userAssignedIdentities/([0-9a-zA-Z-]+)$`)
```
Yes. The biggest reason: `ParseResourceID` is far too lenient for my taste. There are various oddities, but to give you an example:

`///subscriptions///1020304050607-cafe-8090-a0b0c0d0e0f0///resourceGroups/example-resource-group/providers/Microsoft.ManagedIdentity/userAssignedIdentities/teleport-azure` 

will be happily parsed the same as

`/subscriptions/1020304050607-cafe-8090-a0b0c0d0e0f0/resourceGroups/example-resource-group/providers/Microsoft.ManagedIdentity/userAssignedIdentities/teleport-azure`.

After parsing, there is no indication that one had extra slashes sprinkled in. If we were to accept both, then either we would have to compare them after parsing - which would complicate our RBAC code quite a bit - or we would have to accept that `deny(foo)` may not work as expected simply because someone thrown in an extra slash somewhere.

In general, it tightens the logic by requiring a stricter format, which ultimately leads to more secure code as some classes of errors are eliminated.

Still, I think it is worth documenting this and testing that our function is stricter indeed: https://github.com/gravitational/teleport/pull/18675/commits/84831de0badfb8ad8a85015e838e29f9598f66d3

Tener committed:
Remove unused function.

Tener committed:
Add `tsh proxy az` as an alias for `tsh proxy azure`.

Tener committed:
Remove another unused function.

Tener commented:
@greedy52 

> Tested today with an Azure VM. Works as expected.

Thanks!

> Curious is there anyway use the managed identity outside Azure (e.g. on AWS ec2)?

Not without some extra steps, but here is one thing I found on this: https://blog.identitydigest.com/azuread-access-aws/

> Anyway. Overall looks good to me.
> 
> Minor nitpicks:
> 
> 1. It would be great we  can add `--cloud` to `teleport app start`, or maybe update the description for url with an `cloud://` example

I'll add `--cloud` in a follow-up PR.

> 2. Could be just me, that after using `tsh az xxx`, i was expecting the proxy version to be `tsh proxy az` but it's actually `tsh proxy azure`. not sure if worth to have an alias

Good one, I've added an alias.

Tener committed:
Merge branch 'master' into tener/azure-cli

Tener commented:
> @Tener - this PR will require admin approval to merge due to its size. Consider breaking it up into a series smaller changes.

@r0mant can you give it a go?

Tener committed:
Merge branch 'master' into tener/azure-cli

smallinsky commented:
@Tener Can you address:
https://github.com/gravitational/teleport/pull/18675#discussion_r1043361406

Tener committed:
Move app-specific Azure CLI profiles under `~/.tsh/azure`.

Tener committed:
Merge branch 'master' into tener/azure-cli

Tener committed:
Merge branch 'master' into tener/azure-cli

smallinsky reviewed (commented):
@Tener 
Could you take a look at UT fail ? 
```
Logged into Azure app "azure-api".
Your identity: dummy_azure_identity
Example Azure CLI command: tsh az vm list
DEBU [PROXY]     Releasing associated resources - context has been closed. id:305 local:127.0.0.1:24779 login:-teleport-nologin-b22f671c-cec7-4bbe-9a06-52e42acae64e remote:127.0.0.1:36464 teleportUser:alice@example.com srv/monitor.go:266
    azure_test.go:155: 
        	Error Trace:	/__w/teleport/teleport/tool/tsh/azure_test.go:155
        	            				/__w/teleport/teleport/tool/tsh/tsh.go:431
        	            				/__w/teleport/teleport/tool/tsh/azure.go:194
        	            				/__w/teleport/teleport/tool/tsh/azure.go:64
        	            				/__w/teleport/teleport/tool/tsh/tsh.go:1105
        	            				/__w/teleport/teleport/tool/tsh/azure_test.go:66
        	            				/__w/teleport/teleport/tool/tsh/azure_test.go:142
        	Error:      	Not equal: 
        	            	expected: "/user/home/dir/.azure-teleport-azure-api"
        	            	actual  : "/tmp/TestAzure1132987143/001/azure/localhost/azure-api"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-/user/home/dir/.azure-teleport-azure-api
        	            	+/tmp/TestAzure1132987143/001/azure/localhost/azure-api
        	Test:       	TestAzure
```

Tener committed:
Update expected path in a test.

r0mant reviewed (approved):
Bot.

Tener commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -75,6 +75,9 @@ message UserMetadata {
   string AWSRoleARN = 4 [(gogoproto.jsontag) = "aws_role_arn,omitempty"];
 
+  // AzureIdentity is the Azure identity to assume when accessing Azure API.
+  string AzureIdentity = 6 [(gogoproto.jsontag) = "azure_identity,omitempty"];
```
Agreed - in fact, I have that change already scheduled to be made in the upcoming GCP PR. Since reordering fields is a safe operation, I'll do it then.

Thanks for the review!

Tener committed:
Merge branch 'master' into tener/azure-cli

Tener committed:
Merge branch 'master' into tener/azure-cli
